internal/datacoord/meta.go

```diff
@@ -1418,38 +1439,157 @@ func (m *meta) GetRealSegmentsForChannel(channel string) []*SegmentInfo {
 	m.segMu.Lock()
 	defer m.segMu.Unlock()
-	curSegInfo := m.segments.GetSegment(segmentID)
-	if curSegInfo == nil {
-		// TODO: Error handling.
-		log.Ctx(m.ctx).Error("meta update: add allocation failed - segment not found", zap.Int64("segmentID", segmentID))
-		return errors.New("meta update: add allocation failed - segment not found")
-	}
-	// As we use global segment lastExpire to guarantee data correctness after restart
-	// there is no need to persist allocation to meta store, only update allocation in-memory meta.
-	m.segments.AddAllocation(segmentID, allocation)
-	log.Ctx(m.ctx).Info("meta update: add allocation - complete", zap.Int64("segmentID", segmentID))
+
+	segmentsMap := m.getSegmentsMap()
+
+	if segmentsMap == nil {
+		return
+	}
+
+	if segment, ok := segmentsMap[segmentID]; ok {
+		// Get the current LastExpireTime safely
+		currentExpireTime := segment.GetLastExpireTime()
+
+		// Only update if the new time is strictly greater (Monotonically Increasing)
+		if expireTime > currentExpireTime {
+			// Apply the option to clone and set the new maximum time.
+			segmentsMap[segmentID] = segment.ShadowClone(SetExpireTime(expireTime))
+		}
+	}
+}
+
+// AddAllocation adds a new Allocation to the segment's list in a thread-safe manner.
+// This function handles the initialization of the list if it doesn't exist.
+func (m *meta) AddAllocation(segmentID UniqueID, alloc *Allocation) error {
+	// 1. Get the list pointer (RLock for map lookup).
+	m.allocationMu.RLock()
+	list := m.allocations[segmentID]
+	m.allocationMu.RUnlock()
+
+	if list == nil {
+		// 2. The list doesn't exist, initialize it (Write Lock for map modification).
+		m.allocationMu.Lock()
+		// Double-check locking pattern to avoid redundant creation
+		if list = m.allocations[segmentID]; list == nil {
+			list = NewLockedAllocationList()
+			m.allocations[segmentID] = list
+		}
+		m.allocationMu.Unlock()
+	}
+
+	// 3. Acquire Write Lock for the list content and append the new allocation.
+	list.Lock()
+	defer list.Unlock()
+
+	list.allocations = append(list.allocations, alloc)
+
 	return nil
 }
 
-func (m *meta) SetRowCount(segmentID UniqueID, rowCount int64) {
-	m.segMu.Lock()
-	defer m.segMu.Unlock()
-	m.segments.SetRowCount(segmentID, rowCount)
+// GetAllocations retrieves all active Allocations for a given segmentID.
+// It returns a copy of the internal slice to maintain thread safety and prevent external mutation.
+func (m *meta) GetAllocations(segmentID UniqueID) []*Allocation {
+	// 1. Get the list pointer (RLock for map lookup).
+	m.allocationMu.RLock()
+	list := m.allocations[segmentID]
+	m.allocationMu.RUnlock()
+
+	if list == nil {
+		return nil
+	}
+
+	// 2. Acquire Read Lock for the list content (allows concurrent reads).
+	list.RLock()
+	defer list.RUnlock()
+
+	// 3. Create a copy to prevent external modification of the internal slice.
+	result := make([]*Allocation, len(list.allocations))
+	copy(result, list.allocations)
```

Comment:
We need to hold channel lock by assignment.
Which means there should not be any conflicts while doing assignment?

Is it nessesary to copy? I see a allocationPool here, which means the allocation should be very frequent, copy can make this a bottle neck. 

Also if allocation can be modified without channel lock, there might be more serious issues